### PR TITLE
[COD-571] fix: refactored API calls to make them tolerate network failures and retry

### PR DESCRIPTION
**Pull request by Arvi3d**
https://snyksec.atlassian.net/browse/COD-571

- [ ] Ready for review
- [ ] Linked to Jira ticket

### CODE REVIEW

File: src/needle.ts [hunk 1]
```diff
@@ -72,35 +84,39 @@ export async function makeRequest(payload: Payload): Promise<{ success: boolean;
...
+    // console.log(`attempts: ${attempts} && errorCode: ${errorCode}`);
+    if (attempts > 1 && [
+      ErrorCodes.serviceUnavailable,
+      ErrorCodes.badGateway,
+      ErrorCodes.connectionRefused,
+      ErrorCodes.timeout,
+      ErrorCodes.dnsNotFound,
+    ].includes(errorCode)) {
+      attempts--;
+      await sleep(REQUEST_RETRY_DELAY);
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/needle.ts:113]: i'm open to discuss why. i don't don't see it

File: src/needle.ts [hunk 2]
```diff
@@ -72,35 +84,39 @@ export async function makeRequest(payload: Payload): Promise<{ success: boolean;
...
-      emitter.apiRequestLog(`Request error --> ${err}`.slice(0, 399));
-      throw err;
+      errorCode = NETWORK_ERRORS[err.code || err.errno];
+      emitter.apiRequestLog(`Requested url --> ${url} | error --> ${err}`);
     }
+
+    errorCode = errorCode ?? ErrorCodes.serviceUnavailable;
+
+    // Try to avoid breaking requests due to temporary network errors
+    // console.log(`attempts: ${attempts} && errorCode: ${errorCode}`);
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/needle.ts]: cleaned

File: tests/analysis.spec.ts [hunk 3]
```diff
@@ -98,15 +98,15 @@ describe('Functional test of analysis', () => {
...
+              files: 6,
               isSupported: true,
               lang: 'JavaScript',
             },
-            {
-              files: 1,
-              isSupported: true,
-              lang: 'JSX',
-            },
+            // {
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply tests/analysis.spec.ts]: cleaned

**Comment by Arvi3d** [comment]
tests are adjusted and added a test for network failure

File: tests/api.spec.ts [hunk 4]
```diff
@@ -290,15 +296,15 @@ describe('Requests to public API', () => {
...
+              files: 6,
               isSupported: true,
               lang: 'JavaScript',
             },
-            {
-              files: 1,
-              isSupported: true,
-              lang: 'JSX',
-            },
+            // {
```
[the remark this answers is not in the document]
Comment (pkey) [reply tests/api.spec.ts:313]: nit: still here

File: tests/api.spec.ts [hunk 5]
```diff
@@ -274,8 +290,8 @@ describe('Requests to public API', () => {
       if (response.type === 'error') return;
       expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();
 
-      if (response.value.status === AnalysisStatus.complete && response.value.type === 'sarif' ) {
-        expect(response.value.sarif.runs[0].results).toHaveLength(12);
+      if (response.value.status === AnalysisStatus.complete && response.value.type === 'sarif') {
+        expect(response.value.sarif.runs[0].results).toHaveLength(17);
```
Comment (pkey) [inline tests/api.spec.ts:294]: question: I wonder why do we have a different length here? Shouldn't these changes not impact the result?

**Review by pkey — COMMENTED** [review]
looks good, I just wonder why do we have different numbers in the results now? 🤔

File: tests/api.spec.ts (hunk 5 above)
Comment (Arvi3d) [reply tests/api.spec.ts:294]: we have to constantly adjust tests to changes in the engine. 
Newer version of the engine finds more issues in our test repo.

Comment (Arvi3d) [reply tests/api.spec.ts:294]: let me create a separate PR for these adjustments

Comment (Arvi3d) [reply tests/api.spec.ts:294]: done: https://github.com/snyk/code-client/pull/115

File: package.json [hunk 6]
```diff
@@ -49,7 +49,6 @@
     "@typescript-eslint/eslint-plugin": "^4.31.2",
     "@typescript-eslint/parser": "^4.0.1",
     "eslint": "^7.32.0",
-    "eslint-config-airbnb-base": "^14.2.1",
```
Comment (Arvi3d) [inline package.json:52]: we don't really need this plus there is a conflict in dependency versions because of this

**Review by pkey — APPROVED** [review]

**Pull request merged**